### ISSUE TITLE
Add onMapViewFinishedDragging Callback

### ISSUE THREAD

**Issue by Tvax** [osmdroid/osmdroid#1403]
## Issue Type

[ ] Question
[ ] Bug
[x] Improvement
[ ] Build system related
[ ] Performance
[ ] Documentation

## Description and/or steps/code to reproduce the problem
Add a callback that is called when the map finished moving. 

Exemple: the user drag the map from London to Berlin, the user stop dragging the map when it's on Berlin. This callback is called:
```
onMapViewFinishedDragging(BoundingBox boundingBox)
```
or
```
onMapViewFinishedDragging(GeoPoint centerPoint)
```
Then we can use this information to load point of interests only in this area (BoundingBox) or a radius around the center point (GeoPoint).

**Comment by monsieurtanuki** [comment]
I think this should be something simple like "isMapMoving" / "onKeyUp" / "onFlingStopped".
Then you can easily get the map center.

**Comment by Tvax** [comment]
Yes it could be !

**Comment by monsieurtanuki** [comment]
@Tvax @InI4 Would something like that work for you?
````java
mMapView.getOverlays().add(new Overlay() {
	@Override
	public boolean onTouchEvent(MotionEvent event, MapView mapView) {
		if (event.getAction() == MotionEvent.ACTION_UP) {
			Toast.makeText(getContext(), "Current center position is " + mMapView.getMapCenter(), Toast.LENGTH_SHORT).show();
			return true;
		}
		return false;
	}
});
````

**Comment by Tvax** [comment]
Correct me if I'm wrong but the callback will be triggered when the user releases his finger from the screen, not when the map stop moving.
But yeah that would work I guess, if you can get the velocity and direction of the map when the user stops touching the screen (thanks to your callback), and then calculate where the map will end up with the velocity and the direction.

**Comment by monsieurtanuki** [comment]
> Correct me if I'm wrong but the callback will be triggered when the user releases his finger from the screen, not when the map stop moving

Exactly. The only difference I see is when the map is "flinging".
Is that what's bothering you? Then you need something like a "justStoppedFlinging" callback, am I right?

**Comment by Tvax** [comment]
Yes you're totally right !

**Comment by monsieurtanuki** [comment]
Still working on it.
The problem is that it's not that easy to detect when a fling stops. But let's rather talk about Android dev ;)

In `MapView.MapViewGestureDetectorListener.onFling`, there's a call to `Scroller.fling` that works as expected (a visual fling, that eventually stops by a final call to `MapView.onScroll` that checks that the fling is finished).

But sometimes not: the standard case is at the end of a map pan (not systematic).
`onFling` is unexpectedly called, then `Scroller.fling` that here doesn't do anything visually  (nothing happens) and it doesn't call `MapView.onScroll` to say that it's done now. But it waits 3 seconds to say it's done...

My current best idea would be to detect the call to `Scroller.fling` (good cases and bad cases), then to wait "some time" (like 1/10th second): if there hasn't been a call to a map scroll by then, it was a "false" fling.

**Comment by monsieurtanuki** [comment]
Considering that we cannot be 100% sure about when exactly a fling stops, I suggest the following algorithm:
1. set-up a duration in milliseconds (e.g. 200 millis)
1. being called everytime the end-user stopped panning the screen (basically, MotionEvent.UP and no map move in the next _n_ millis)
1. being called everytime a fling starts and the map didn't move in the next _n_ millis (in order to bypass the fling bug where nothing happens during 3 seconds)
1. begin called everytime a fling stops and no map move in the next _n_ millis (and actual map moves in the last _n_ millis, in order to discard the fling-start-waiting-3-seconds case)
1. possibly include `MapController` animations in the process in a second step

@Tvax @InI4 What do you guys think of that?

**Comment by InI4** [comment]
Explaining it this way, it looks there could be` onMapViewFinishedDragging()` and `onMapViewFinishedMoving()` which are actually different events in case of flings. 

- I am not sure I understand what the duration is, is it "n" below?
- You mention an exact way to determine, a drag is over: Map moved, now UP.
- Isn't there anything exact with fling, couldn't n be something like just 2 UI ticks?
- I like the idea 5.

**Comment by monsieurtanuki** [comment]
Correct: the _n_ means a time duration in millis. We need a (tiny) amount of time to check if the fling stopped, because there's (unfortunately) no bullet-proof "fling's stopped" event.
It may also make sense UX-wise if we wait a little before we display a popup to the end-user about the new current POIs if the end-user's just paused 100ms before going on panning the map.

I'm afraid I cannot see the difference you make between dragging and moving. In the case of the OP, I believe the basic need is to detect that the map's just stopped moving, which basically happens if:
* the user's stopped panning the map and doesn't touch the screen anymore
* the map's stopped flinging
* (my idea "number 5") the map's stopped moving from a map controller animation, a zoom in for instance

**Comment by Tvax** [comment]
Sorry for the late response.

Can't we "fix" the 3s delay call to the `MapView.onScroll` method, that would be a cleaner solution I guess, nonetheless I have no idea if this will have an impact on the rest of the SDK.

Otherwise I think adding a timer would be a great idea to bypass the problem!
